wasmplugin/plugin.go

```diff
@@ -187,41 +191,63 @@ func (ctx *httpContext) OnHttpRequestBody(bodySize int, endOfStream bool) types.
 	// Do not perform any action related to request body if SecRequestBodyAccess is set to false
 	if !tx.IsRequestBodyAccessible() {
 		proxywasm.LogDebug("skipping request body inspection, SecRequestBodyAccess is off.")
+		ctx.processedRequestBody = true
+		interruption, err := tx.ProcessRequestBody()
+		if err != nil {
+			proxywasm.LogCriticalf("failed to process request body: %v", err)
+			return types.ActionContinue
+		}
+
+		if interruption != nil {
+			return ctx.handleInterruption("http_request_body", interruption)
+		}
+
 		return types.ActionContinue
 	}
 
-	ctx.requestBodySize += bodySize
-	// Wait until we see the entire body. It has to be buffered in order to check that it is fully legit
-	// before sending it upstream
-	if !endOfStream {
-		return types.ActionPause
-	}
+	if bodySize > 0 {
+		b, err := proxywasm.GetHttpRequestBody(ctx.bodyReadIndex, bodySize)
+		if err == nil {
+			interruption, _, err := tx.WriteRequestBody(b)
+			if err != nil {
+				proxywasm.LogCriticalf("failed to write request body: %v", err)
+				return types.ActionContinue
+			}
 
-	if ctx.requestBodySize > 0 {
-		body, err := proxywasm.GetHttpRequestBody(0, ctx.requestBodySize)
-		if err != nil {
-			proxywasm.LogCriticalf("failed to get request body: %v", err)
+			if interruption != nil {
+				return ctx.handleInterruption("http_request_body", interruption)
+			}
+
+			ctx.bodyReadIndex += bodySize
```

Comment:
Some of the logs produced by FTW when adding `proxywasm.LogDebugf(fmt.Sprintf("[%d] index: %d, bodySize: %d, EOS: %t", ctx.contextID, ctx.bodyReadIndex, bodySize, endOfStream))` to the very beginning of the `OnHttpRequestBody` callback are:

```
[2023-01-23 14:22:05.150338][27][debug][wasm] [source/extensions/common/wasm/context.cc:1167] wasm log coraza-filter my_vm_id: [3] index: 0, bodySize: 292, EOS: false
[2023-01-23 14:22:05.150347][27][debug][wasm] [source/extensions/common/wasm/context.cc:1167] wasm log coraza-filter my_vm_id: [3] index: 292, bodySize: 292, EOS: true
```

which is related to [this payload](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920410.yaml#L903)

and 

```
[2023-01-23 14:22:05.133997][21][debug][wasm] [source/extensions/common/wasm/context.cc:1167] wasm log coraza-filter my_vm_id: [3] index: 0, bodySize: 67372, EOS: false
[2023-01-23 14:22:05.134087][21][debug][wasm] [source/extensions/common/wasm/context.cc:1167] wasm log coraza-filter my_vm_id: [3] index: 67372, bodySize: 67372, EOS: true
```

related to [this payload](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920410.yaml#L20)

whereas logs produced in tests are:

```
2023/01/24 01:23:12 proxy_debug_log: [3] index: 0, bodySize: 5, EOS: false
2023/01/24 01:23:12 proxy_debug_log: [3] index: 5, bodySize: 5, EOS: false
2023/01/24 01:23:12 proxy_debug_log: [3] index: 10, bodySize: 5, EOS: false
2023/01/24 01:23:12 proxy_debug_log: [3] index: 15, bodySize: 5, EOS: false
2023/01/24 01:23:12 proxy_debug_log: [3] index: 20, bodySize: 5, EOS: false
2023/01/24 01:23:12 proxy_debug_log: [3] index: 25, bodySize: 5, EOS: false
2023/01/24 01:23:12 proxy_debug_log: [3] index: 30, bodySize: 2, EOS: true
```

related to the payload `animal=bear&food=honey&name=pooh` which is 32 char length.

cc @anuraaga @fzipi